ow_power_system/src/power_system_node.cpp

```diff
@@ -113,39 +75,94 @@ int main(int argc, char* argv[]) {
 
   ROS_INFO_STREAM_NAMED("power_system_node", "power_update_rate is set to: " << power_update_rate << " Hz");
   
-  // Initialize time step to 0
-  int t_step = 0;
-
-  //set our indices
-  int time_i = 0;
-  int voltage_i = 1;
-  int rul_i = 2;
-  
   // ROS Loop. Note that once this loop starts,
   // this function (and node) is terminated with an interrupt.
   
   ros::Rate rate(power_update_rate);
   //individual soc_msg to be published by SOC_pub
   std_msgs::Float64 soc_msg;
-  std_msgs::Float64 rul_msg;
+  std_msgs::Int16 rul_msg;
+  std_msgs::Float64 tempbat_msg;
   ROS_INFO ("Power system node running");
 
-  int power_lifetime = power_csv.size();
+  int count = 0;
+  
   while (ros::ok()) {
-    if (t_step < power_lifetime) {
-      soc_msg.data=power_csv[t_step][voltage_i];
-      rul_msg.data=power_csv[t_step][rul_i];
-    } //publishes last voltage value if sim time exceeds the length of the csv
-
-    //publish current SOC
-    SOC_pub.publish(soc_msg);
-    RUL_pub.publish(rul_msg);
-    t_step+=1;
-    
-    ros::spinOnce();
-    rate.sleep();
-   
-  }
 
+    // For each line of data in the example file, run a single prediction step.
+    for (const auto & line : power_data) {
+      // Get a new prediction
+      auto prediction = prognoser->step(line);
+      // Get the event for battery EoD. The first line of data is used to initialize the observer,
+      // so the first prediction won't have any events.
+      if (prediction.getEvents().size() == 0) {
+          continue;
+      }
+      auto eod_event = prediction.getEvents().front();
+      // The time of event is a `UData` structure, which represents a data
+      // point while maintaining uncertainty. For the MonteCarlo predictor
+      // used by this example, the uncertainty is captured by storing the
+      // result of each particle used in the prediction.
+      UData eod_time = eod_event.getTOE();
+      if (eod_time.uncertainty() != UType::Samples) {
+        ROS_WARN_NAMED("power_system_node", "Unexpected uncertainty type for EoD prediction");
+        if (count > 0) {
+          RUL_pub.publish(rul_msg);
+          //SOC_pub.publish(soc_msg);
+          //TempBat_pub.publish(tempbat_msg);
+        } else {
+          rul_msg.data = 0;
+          //soc_msg.data = 0.0;
+          //TempBat_msg.data = 0.0;
+          RUL_pub.publish(rul_msg);
+          //SOC_pub.publish(soc_msg);
+          //TempBat_pub.publish(tempbat_msg);
+        }
+        ros::spinOnce();
+        rate.sleep();
+        continue;
+      }
+      
+      // For this example, we will print the median EoD.
+      auto samples = eod_time.getVec();
+      std::sort(samples.begin(), samples.end());
```

Comment:
Paraphrasing here, this can be simplified as follows (this code assumes you have set `rul_msg.data = 0` on the start:
```suggestion
      UData eod_time = eod_event.getTOE();
      if (eod_time.uncertainty() != UType::Samples) {
        // Log warning and don't update the last value
        ROS_WARN_NAMED("power_system_node", "Unexpected uncertainty type for EoD prediction");
      } else { // valid prediction
        // For this example, we will print the median EoD.
        auto samples = eod_time.getVec();
        std::sort(samples.begin(), samples.end());
        ...
        rul_msg.data = rul_median;
      }
```